File: Android-Client/app/src/main/java/com/grabtube/android/domain/usecase/ScanQRCodeUseCase.kt
package com.grabtube.android.domain.usecase

import com.grabtube.android.domain.model.QRScanResult
import com.grabtube.android.domain.repository.QRScannerRepository
import javax.inject.Inject

class ScanQRCodeUseCase @Inject constructor(
    private val repository: QRScannerRepository
) {
    suspend operator fun invoke(): Result<QRScanResult> {
        val hasPermission = repository.hasCameraPermission()
        if (!hasPermission) {
            val permissionResult = repository.requestCameraPermission()
            if (permissionResult.isFailure) {
                return Result.failure(permissionResult.exceptionOrNull() ?: Exception("Permission denied"))
            }
        }

        return repository.scanQRCode()
    }
}

File: Flutter-Client/lib/domain/usecases/scan_qr_code_usecase.dart
import 'package:dartz/dartz.dart';
import '../entities/qr_scan_result.dart';
import '../repositories/qr_scanner_repository.dart';

/// Use case for scanning QR codes from camera
class ScanQRCodeUseCase {
  final QRScannerRepository _repository;

  ScanQRCodeUseCase(this._repository);

  Future<Either<String, QRScanResult>> call() async {
    try {
      // Check if camera permission is granted
      final hasPermission = await _repository.hasCameraPermission();

      if (!hasPermission) {
        // Request permission if not granted
        final permissionResult = await _repository.requestCameraPermission();

        // Check if permission request failed
        if (permissionResult.isLeft()) {
          return permissionResult.fold(
            (error) => Left(error),
            (_) => Left('Permission denied'),
          );
        }
      }

      // Scan QR code
      final result = await _repository.scanQRCode();

      return result.fold(
        (error) => Left(error),
        (scanResult) async {
          // Save to history
          await _repository.saveScanResult(scanResult);
          return Right(scanResult);
        },
      );
    } catch (e) {
      return Left('Failed to scan QR code: ${e.toString()}');
    }
  }
}
